text/0000-named-tracers.md

```diff
@@ -0,0 +1,62 @@
+# Named Tracers 
+
+_Creating tracers using a factory mechanism and naming those tracers in accordance with the library that provides the instrumentation for a traced component._
+
+## Suggested reading
+
+* [Proposal: Tracer Components](https://github.com/open-telemetry/opentelemetry-specification/issues/10)
+* [Global Instance discussions](https://github.com/open-telemetry/opentelemetry-specification/labels/global%20instance)
+
+## Motivation
+
+Instead of having a global tracer singleton, every instrumentation library needs to create its own tracer, which also has a proper name that identifies the library / logical component that uses this tracer. If there are two different instrumentation libraries for the same technology (e.g. MongoDb), these instrumentation libraries should have distinct names.
+
+This proposal originates from an `opentelemetry-specification` proposal on [components](https://github.com/open-telemetry/opentelemetry-specification/issues/10), but the term "components" was quickly found to being too generic and possibly misleading, since it is already used frequently in differing contexts and connotations.
+
+The purpose of _Named Tracers_ is the ability to allow for a more fine-grained control of the amount of tracing data being generated by a library or application. Associating a name (e.g. _"io.opentelemetry.contrib.mongodb"_) with a tracer allows for a logical coupling of this tracer and a certain technology or library (like _MongoDB_ in this example).
+
+By using *Named Tracers* a library can enable / disable tracing of a specific library, which might be necessary in case:
+* A library causes too much runtime or memory overhead.
+* A library has issues or bugs (e.g. does not call `span.end()` in every case).
+* "Double-Tracing" becomes a problem. This can be the case if a vendor (e.g. Dynatrace) might already have built-in auto-instrumentation for a certain library. In this case the vendor might want to decide (automatically or configuration-based) which of the implementations should be enabled or disabled.
```

Comment:
This third issue is the only one that really moves me as a problem-worth-solving, at least at this stage of the project... bugs are bugs and should be fixed; and overhead can be controlled without introducing something this complex and general-purpose.

If we're focused on the double-instrumentation problem, I would suggest doing something simpler, though – it seems like we want more of a "registry" of the application's telemetry surface area where we can ensure single-coverage / conflict resolution / etc. I am very much in favor of coming up with a credible (though I'd also argue for "minimal" / simple) solution to that problem, but I don't think it would be "named tracers" per se.

Reply:
We've seen demand from various enterprise customers to quickly turn off specific instrumentation also for reasons (1) and (2) under certain circumstances. An example are cases when the instrumentation usually works well but misbehaves in certain use cases not considered when adding the instrumentation, e.g., when an instrumented library is integrated in a new, high-load environment.

To me, the registry you are proposing sounds like something that would be build up from the names passed to the TracerFactory proposed in this RFC. How else would you expect such a registry to be populated? Or are you thinking of a centrally maintained global registry of all known instrumentations hosted by OpenTelemetry rather than one dynamically built up for each specific application instance?

Reply:
Fully agree to @arminru 's comment.

I'll try to outline a scenario that explain cases (1) and (2):

Customer runs an app in production. It's monitored by a certain APM vendor (e.g. Dynatrace). Everything works smoothly.

Customer deploys a new version of the application, which has new OpenTelemetry instrumentation (either new library with OTel support or custom instrumentation). However, the new instrumentation is far too chatty (produces too many spans), is too heavyweight (e.g. while populating attributes, it queries expensive API's) or is buggy (span.end not properly called). In test and pre-prod this is maybe not really a problem, because only the No-Op Tracer is tested.

As soon as that deployment hits production, a different Tracer is used and suddenly the problems surface. In that case a customer would turn to the APM vendor for support and in enterprise scenarios a quick solution is needed. Customers are often unable to quickly deploy new application versions, so they would end up being forced to disable APM monitoring until then.

In such a case, we would like to be able to identify various active "instrumentations" and disable the offending one. That way, we could troubleshoot "live", without any application re-deployment.

Reply:
@arminru @discostu105 thanks for your replies above. Much of this makes sense to me.

It sounds like you want to create a name – and a "single throat to choke" – for some set of instrumentation (that likely pertains to a single library / component / whatever).

After sleeping on this, I think the naming here is confusing me... I think of a "Tracer" as basically a "sink" for telemetry data; where here the term "Tracer" is actually being used to describe a "source" for telemetry data.

I stand by my desire for a (simple, basic) registry API, and I would also suggest calling these something other than a "Tracer" since this new concept is used to describe where telemetry comes *from*, not where it goes *to*.

Thoughts??

Reply:
@bhs Thanks for your reply. We're totally open for a different way of achieving this goal. Do you have a suggestion for such a registry API?

In this proposal, the `TracerFactory` basically acts as such a registry. IMHO it is a very simple approach and it's borrowed from Logging (log4j), where this you'd also create a Logger for each "source". E.g.

```
/* Get the logger for the actual class name to be printed on */
private final Logger log = LoggerFactory.getLogger(App.class);
```

Reply:
I like the `log4j` model as well. This suggests that the SDK concept we've been discussing is really a factory for named Tracer and Meter interfaces, and the calling convention will be to statically allocate one per module of code, using the global SDK I presume:

```
import io.opentelemetry.GlobalSDK;

  private final Tracer tracer = GlobalSDK.getTracer(MyClass.class);
  private final Meter meter = GlobalSDK.getMeter(MyClass.class);
```

Reply:
TL;DR I like @jmacd's proposal ^^^

Reply:
+1 for the factory like method.

@jmacd I think the factory should be part of the main trace/metrics/etc package and the global instance should have a `getTracerFactory().getTracer("my_component");`. The reason is that if people want to manually pass a `Tracer` to a plugin for example they should pass the `TracerFactory` not the initialized `Tracer` with component name.

Maybe in future we would like to have a "telemetry level" similar with logs.

Reply:
Same here, like @jmacd's proposal.